include/cpp-statsd-client/StatsdClient.hpp

```diff
@@ -116,33 +132,52 @@ inline void StatsdClient::send(const std::string& key,
                                const int value,
                                const std::string& type,
                                const float frequency) const noexcept {
-    const auto isFrequencyOne = [](const float frequency) noexcept {
-        constexpr float epsilon{0.0001f};
-        return std::fabs(frequency - 1.0f) < epsilon;
-    };
-
-    // Test if one should send or not, according to the frequency rate
-    if (!isFrequencyOne(frequency)) {
-        if (frequency < static_cast<float>(std::rand()) / RAND_MAX) {
-            return;
-        }
+    constexpr float epsilon{0.0001f};
+    const bool isFrequencyOne = std::fabs(frequency - 1.0f) < epsilon;
+
+    // If you are sampling at a rate less than 1 (ie not sending every metric) and the RNG is above the sampling rate
+    // then we don't need to send this metric this time
+    if (!isFrequencyOne && (frequency < std::uniform_real_distribution<float>(0.f, 1.f)(m_random_engine))) {
+        return;
     }
 
-    // Prepare the buffer, with a sampling rate if specified different from 1.0f
-    char buffer[256];
-    if (isFrequencyOne(frequency)) {
+    // Prepare the buffer and include the sampling rate if it's not 1.f
+    std::string buffer(256, '\0');
+    int string_len = -1;
+    if (isFrequencyOne) {
         // Sampling rate is 1.0f, no need to specify it
-        std::snprintf(buffer, sizeof(buffer), "%s%s:%d|%s", m_prefix.c_str(), key.c_str(), value, type.c_str());
+        string_len = std::snprintf(
+            &buffer.front(), buffer.size(), "%s%s:%d|%s", m_prefix.c_str(), key.c_str(), value, type.c_str());
     } else {
         // Sampling rate is different from 1.0f, hence specify it
-        std::snprintf(
-            buffer, sizeof(buffer), "%s%s:%d|%s|@%.2f", m_prefix.c_str(), key.c_str(), value, type.c_str(), frequency);
+        string_len = std::snprintf(&buffer.front(),
+                                          buffer.size(),
+                                          "%s%s:%d|%s|@%.2f",
+                                          m_prefix.c_str(),
+                                          key.c_str(),
+                                          value,
+                                          type.c_str(),
+                                          frequency);
+    }
+
+    // A valid metric must be at least 6 chars in length
+    if(string_len < 6){
+        //TODO: we dont have access to the error message here
+        return;
     }
 
+    // Trim the trailing null chars
+    // TODO: perf improvement: send the len along and keep the buffer around as a member variable
+    buffer.resize(std::min(string_len, static_cast<int>(buffer.size())));
```

Comment:
so i noticed that we were copying a `char[]` into an `std::string` when we call `send` below. i did a small refactor to just not copy but rather start with a string in the first place. long term i think we should make the buffer a member variable and keep it a fixed size and then send it and its `snprintf`'d length to the send funciton so that it can avoid copying all together (except in the case it has to be batched).

i want to make another pr where refactor the batching a bit to maybe keep a single buffer with all the unsent stuff in it (either via the background thread or not) and just flush that from time to time. this would avoid any unneeded reallocs or deallocs (except when we get over the UDP limit which is like 64k)